server/vitals.js

```diff
@@ -159,21 +185,44 @@ async function checkButtonHeartbeat() {
             const logMessage = `Disconnection: ${client.displayName} ${button.displayName} Button delay is ${buttonDelay} seconds.`
             helpers.logSentry(logMessage)
 
+            // Check for disconnected gateways - if any disconnected gateways are returned, do not message the client
+            const gateways = await db.getDisconnectedGatewaysWithClient(client)
+            if (gateways !== null && gateways.length === 0) {
+              // Store the client info
+              if (!clientMessages[client.id]) {
+                clientMessages[client.id] = {
+                  client,
+                  disconnectedButtons: [],
+                  reconnectedButtons: [],
+                }
+              }
+              // Store the disconnected button name
+              clientMessages[client.id].disconnectedButtons.push(button.displayName)
+            }
             await db.updateButtonsSentVitalsAlerts(button.id, true)
-
-            // TODO Also send a text message to Responders and Heartbeat Phone Numbers once we know that these messages are reliable
           }
           // TODO Also send a text message reminder once we know that these messages are reliable
         } else if (button.sentVitalsAlertAt !== null) {
           const logMessage = `Reconnection: ${client.displayName} ${button.displayName} Button.`
           helpers.logSentry(logMessage)
 
+          // Store the client info
+          if (!clientMessages[client.id]) {
+            clientMessages[client.id] = {
+              client,
+              disconnectedButtons: [],
+              reconnectedButtons: [],
+            }
+          }
+          // Store the reconnected button name
+          clientMessages[client.id].reconnectedButtons.push(button.displayName)
           await db.updateButtonsSentVitalsAlerts(button.id, false)
-
-          // TODO Also send a text message to Responders and Heartbeat Phone Numbers once we know that these messages are reliable
         }
       }
     }
+
+    // Log one message per client
+    logClientMessage(clientMessages)
```

Comment:
Yay! I love how the functionality of logging the messages for every client comes down to this one function and one argument. It's a really clean solution; well done!

Reply:
Thanks! :)